### ISSUE TITLE
Feature request: Graceful shutdown

### ISSUE THREAD

**Issue by whitslack** [ElementsProject/lightning#4842]
@ZmnSCPxj [explained](https://github.com/ElementsProject/lightning/issues/4824#issuecomment-934235242):

> HTLCs being in-flight are a normal part of operation and that state can persist for _days_; we cannot delay daemon shutdown until there are no in-flight HTLCs when in-flight HTLCs can persist for days at a time.

It is certainly true that daemon shutdown cannot be deferred until all HTLCs are cleared, but a significant improvement in Lightning UX could be achieved by implementing a **graceful shutdown** that would block the addition of new HTLCs and wait for up to a specified timeout for all HTLCs to clear.

### Feature Request
* Add an optional `timeout` parameter to the `stop` RPC to specify a graceful shutdown timeout.
* When `stop` begins executing, begin refusing all requests to add HTLCs to channels (both from peers and from local commands).
* When there are no in-flight HTLCs, terminate the daemon.
* When the timeout expires, terminate the daemon.

Implementing this feature would reduce the occurrence of slow payment attempts for users of the Lightning Network. Rebooting a C-Lightning server can take several minutes, during which time any users with in-flight HTLCs must wait. This is bad UX and is not helping Lightning adoption. We can't easily fix HTLCs that go out to lunch and never return, but we *can* avoid dropping fresh HTLCs on the floor while *we* go out to lunch.

**Comment by whitslack** [comment]
*N.B.:* @ajpwahqgbi has attempted to write a [C-Lightning graceful shutdown plugin](https://github.com/ajpwahqgbi/lightning-tools/tree/master/clightning-plugins/gracefulshutdown), but it doesn't work correctly, as it apparently blocks the daemon's event loop and prevents any HTLCs from clearing out at all, thus actually exacerbating the problem it's trying to remedy. Also, it lacks a timeout.

**Comment by ZmnSCPxj** [comment]
This may not be (perfectly) possible in terms of the BOLT protocol.

In BOLT#2, there is an implicit ordering requirement. Suppose you are doing (or want to do, ha) a graceful shutdown, and I am your peer.  There is an HTLC you already offered to me.  Now, suppose I send the following messages to you:

* `update_add_htlc` of a brand-new HTLC.
* `update_fulfill_htlc` of the previous HTLC you offerred to me.

You can either respond:

* To neither message.
* To the first message.
* To the first and second message.

However, note that the first message *adds* an HTLC, which is further from your goal of reducing the number of HTLCs in-flight before you finalize shutdown.

You **cannot** respond to just the second message and remove an HTLC without adding the first message HTLC.

The first option (respond to neither message) is equivalent to just shutting down with a `SIGKILL` immediately then and there instead of trying for a timed-out graceful shutdown.  The third option is not  a marked improvement: you remove one HTLC but gain another anyway.

I think that properly speaking we may need a spec change, where a node tells its peers "no more `update_add_htlcs` until I reconnect" but will still accept fulfilling and failing HTLCs to reduce in-flight HTLCs.

--

However, we can (maybe?) offer a partial solution.

We can have `channeld` accept a `graceful_shutdown` message, which puts it in `graceful_shutdown` mode.  It should also probably get that flag in its `init` message too.

In this mode, `channeld` will handle `update_fulfill_htlc` and  `update_fail_htlc` as normal.  However, if it receives `update_add_htlc` it will immediately exit, which automatically closes the peer connection and appears to the peer as a sudden shutdown of our node.

This at least lets us have *some* chance to reduce the number of HTLCs without gaining more HTLCs, without having to modify the spec (which would take longer).

**Comment by whitslack** [comment]
@ZmnSCPxj: In your example, why not reply to the `update_add_htlc` with a temporary failure `0x2002` (`temporary_node_failure`) and process the `update_fulfill_htlc` normally? We should endeavor to keep the `channeld` running if there are any HTLCs in flight. The `channeld` processes for channels with no HTLCs in flight can be terminated immediately (i.e., sooner than `lightningd` times out and terminates).

**Comment by ZmnSCPxj** [comment]
You cannot reply to `update_add_htlc` with an routing failure --- that is done with a *later* `update_fail_htlc`.  Unfortunately, that means *accepting* the `update_add_htlc` *now*, then *after* the HTLC has been irrevocably committed (and therefore we have exchanged signatures and added the new HTLC as in-flight), we will then `update_fail_htlc` with the `temporary_node_failure`.  The counterparty might not reply immediately (we have no control over their code) and our graceful shutdown timer might time out before they reply, meaning we "gracefully" shut down with an *additional* HTLC.

So no, your only recourse is to drop the connection so that neither you nor the counterparty ever instantiate the new incoming HTLC.

(In fact I believe we do this now, if we have a just-arrived HTLC that is to be forwarded, but before we actually *have* sent out the `update_add_htlc` to the next hop, and we get shutdown, we will `temporary_node_failure` the HTLC; though @rustyrussell probably knows the HTLC state machine much better than I do.)

**Comment by whitslack** [comment]
Damn. Who the hell designed this protocol? It has so many novice mistakes. Anyway, thanks for your explanations.

**Comment by ZmnSCPxj** [comment]
N00b LN devs 4->5 years ago (all of us were n00bs, LOL).

Mostly the concern then was to ensure that we could always have a fallback safely on-disk before doing potentially funds-losing operations, which is why the protocol was designed that way; graceful shutdowns were not a consideration, but sudden expected I-tripped-over-the-power-cord-sorrry shutdowns *were*, and the protocol was designed so that sudden disconnections are perfectly fine (which is why I suggest just disconnecting, the protocol is designed to survive that).

**Comment by whitslack** [comment]
For what it's worth, I don't mean that responding to protocol messages in order is a mistake. I only meant that having no mechanism for rejecting protocol messages (other than disconnecting) seems like an oversight. Similarly with the whole thing where you can't send an "error" to your channel peer without triggering them to force-close your channel.

**Comment by ZmnSCPxj** [comment]
Well, I suppose I now have to detail it exactly, and the rationale.  You can try reading through [BOLT#2](https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#normal-operation) too.

Basically, I lied when I said your options are to respond to none of the updates, one of them, or both.  In reality, multiple `update_`\* messages are effectively *batched* into a single Poon-Dryja update (i.e. signature followed by revocation).  And it is only the entire batch you can ignore or reject, because it is the signing and revocation which are "the real" events that commit everyone to a new state.

So your only real options, if the counterparty gave you `update_add_htlc` followed by `update_fail_htlc`, are:

* Take the entire batch (which does not reduce the number of HTLCs in flight).
* Ignore the entire batch (which keeps your current number of HTLCs in flight).

You ignore the batch by ignoring their `commitment_signed` and not giving back a `revoke_and_ack`, but in practice most implementations will stall waiting on your `revoke_and_ack` and that is no better than just disconnecting to reject the entire batch (and disconnecting is just plain cleaner and is less likely to trigger edge-case bugs).

The reason to batch is that some of the low-level networking protocols are much more amenable to having as few turnarounds as possible.  Meaning you would prefer to have one side do multiple message sends, then the other side.  For example it might use a shared media (Ethernet, WiFi), or the protocol might have high latency but high bandwidth (Tor, long-distance cable).  For my example it would look like:

    You                         Me
     |                          |
     |<-----update_add_htlc-----|
     |<---update_fulfill_htlc---|
     |<----commitment_signed----|
     |                          |
     |------revoke_and_ack----->|
     |-----commitment_signed--->|
     |                          |
     |<-----revoke_and_ack------|
     |                          |

Now, what you want to do would be something like this:

    You                         Me
     |                          |
     |<-----update_add_htlc-----|
     |                          |
     |-------accept_update----->|
     |                          |
     |<---update_fulfill_htlc---|
     |                          |
     |-------accept_update----->| 
     |                          |
     |<----commitment_signed----|
     |                          |
     |------revoke_and_ack----->|
     |-----commitment_signed--->|
     |                          |
     |<-----revoke_and_ack------|
     |                          |

Because of the need to turn around and accept each individual `update_` message, the latency is worsened; there are now multiple turnarounds, which is bad for shared-media network transports and for transports with high latency.  We could do some amount of batching negotiation, i.e. there could be a "okay that is the batch of changes I want to add" which should be replied with "hmm I only accept these particular changes" before signing and revoking, but that still adds one more turnaround.

And a good part of the slowness in forwarding is due precisely to these turnarounds or roundtrips.  Already with the current design (which is already optimized as well as we can figure out, barring stuff like my weird Fast Forwards idea that you can read about on the ML) you get the 1.5 roundtrips, and that is ***repeated*** at ***each*** hop in a payment route.  Even with a "I only accept these particular changes in that batch" you still add 1 more roundtrip, meaning that is 2.5 roundtrips per hop, on the forwarding of a payment.

It gets worse when a failure happens at a later stage.  Failures need to wait for the outgoing HTLC to be removed before we can consider asking the incoming HTLC to be removed, so the 1.5 roundtrips is *again* ***repeated*** at each hop going back.  If we allow for filtering which messages we accept that rises to 2.5 roundtrips.

And forwarding payments and returning failures happen a *lot* more often than controlled shutdown of forwarding nodes does (because forwarding nodes want to be online as much as possible), so it makes sense to optimize the protocol for that, at the expense of controlled shutdown.  The 1.5 roundtrips are the minimum safe possible if we want to remain robust against *uncontrolled* shutdown, too.

So it seems to me that you want some kind of protocol message to say "okay I am in graceful shutdown mode and will reject batches of updates that increase HTLCs".   The counterparty might already have sent out some updates taht *increase* the number of HTLCs before it gets that message, because latency, so we need a code path to handle that somehow.  Then the counterparty has to ensure that it does not make a batch of updates that increases the number of HTLCs  --- and the easiest way to implement that, with fewer edge cases for bugs in implementation, is to simply stop all updates, which is not much different from, well, disconnecting.

**Comment by whitslack** [comment]
Thank you for all the explanation. I was aware that you'd be committing to HTLCs that I'd subsequently be refusing, but my expectation is that I'd revoke that commitment after you send me a new one that includes only the HTLC that I accepted. That does imply that I'd be revoking two commitments at once, and you would have to assume that *either* of them could be used against you until I give you that revocation.

> while `n` is valid but `n+1` has been revoked.

Is that even possible? I thought the way SHA chains work is that a revocation of a particular state gives you the ability to derive revocations of all preceding states.

Anyway, yes, splitting this into two requests is clearly the way forward. The first half seems trivial to implement correctly, and it would go a *long* way toward reducing the adverse effects of taking down a popular routing node for maintenance.

**Comment by whitslack** [comment]
@ZmnSCPxj: Would the proposed [Quiescence Protocol](https://github.com/lightningnetwork/lightning-rfc/pull/869) be of use here?

**Comment by ZmnSCPxj** [comment]
> Hmm, maybe not, since that protocol, as proposed at least, prevents removing HTLCs as well as adding them. But maybe the proposal could be amended so "add" updates and "remove" updates are independently toggleable?

Correct, not *directly* useful.

The quiescence protocol currently does not specify how the quiescence period ends, and it is intended for changing the commitment protocol (i.e. switching from the current Poon-Dryja to a different variant, or adding an offchain "conversion transaction" to switch to e.g. Decker-Russell-Osuntokun, or maybe just allowing implementations on both sides to mutate their database for an "allow PTLCs" flag safely with as little chance of unexpected stuff happening).  But maybe *bits* of its implementation can be used.

**Comment by whitslack** [comment]
In the case that not all outstanding HTLCs can be resolved before the graceful shutdown timeout expires (and thus the node is forced to go offline with HTLCs still outstanding), it would be nice if CLN would emit a log message informing the node operator of how long they can remain offline before one of their HTLCs will hit its deadline.

**Comment by sangbida** [comment]
This issue is being closed as part of issue cleanup.

**Reason**: About deprecated feature

If you believe this issue should remain open, please comment and we can reopen it.

**Comment by whitslack** [comment]
This feature has not been implemented. In fact there is an open PR to implement it.
